enclave/crypto/openssl/cert.c

```diff
@@ -0,0 +1,479 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include "cert.h"
+#include <ctype.h>
+#include <openenclave/bits/result.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/internal/cert.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/safecrt.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+#include <openssl/x509.h>
+#include <openssl/x509_vfy.h>
+#include <openssl/x509v3.h>
+#include <string.h>
+
+/*
+ * Value used by the _decode_oid_to_str() function. Although the OID
+ * standard does not limit the depth of an OID definition tree (i.e., the
+ * number of arcs), our implementation only supports a simple decoding
+ * with a limited depth (i.e., decoding into a fixed size string).
+ */
+#define OE_OID_MAX_LENGTH 200
+
+/*
+ * Parse the name string into X509_NAME struct. The format of the string is
+ * "KEY1=VALUE1,KEY2=VALUE2,KEY3=VALUE3...". The implementation is based
+ * on the mbedtls_x509_string_to_names from Mbed TLS.
```

Comment:
Instead of cloning an implementation, have we considered simply using the mbedtls implementation directly? It's a nice code base because many files are self contained, and we have precedence for this on the host side with the BCrypt wrappers using mbedtls' implementation of bignum because BCrypt doesn't expose its own. 

From a cursory glance, it looks like this would just pull in x509_create.c and asn1write.c, although the former does have a lot of symbols that are not easy to hide in the resulting .a file (shouldn't be a problem realistically since at link time we specify `-gc-sections` though).

I think that also simplifies the testing story in that we don't need to maintain parser specific tests for our fork of it.

Reply:
Just checked the implementation of `mbedtls_x509_string_to_names`, which is defined as follows.
```
int mbedtls_x509_string_to_names(mbedtls_asn1_named_data **head, const char *name);
```

The case of using the function may not be straightforward given that the function not only parses the string but also sets the Mbed TLS-specific data structure (`mbedtls_asn1_named_data`).

Reply:
I think that's a fair assessment barring any deeper investigations of trying to map the linked list of `mbedtls_asn1_named_data` objects to `ASN1_OBJECT` for `X509_NAME_add_entry_by_OBJ` 

It would probably be helpful to add the implications of what this means for a user in the comments (e.g. unlike the parsing behavior users of OpenSSL CLI might be familiar with, we do not support multi-RDN names with "+" in them.) There's a broader compat question since OpenSSL seems to expect a different input format in their CLI app (e.g. `/type0=value0/type1=value1/type2=...`) and we might want to track that as an open and move on.

Reply:
Added comments around the difference between OpenSSL CLI.